File: code/photodet.py
# === DYNAMICS OF A SYSTEM SUBJECT TO A CONTINOUS PHOTO-DETECTION ===

# import required libraries and objects
import sys
import numpy as np
import random as rnd                                # to deal with pseudorandom numbers
import os                                           # to count the cores
from concurrent.futures import ProcessPoolExecutor  # for parallel computing  
import time                                         # to count the execution time of a process
from my_library.my_system import *                  # we import everything from my_library
from my_library.my_functions import *
from qutip import sigmam, qeye
import pickle   # to print complex objects on a file as they are
import gc       # garbage collector for manual memory cleanup

# command that gives the number of usable physical cores (for an eventual parallel compututing)
num_cores = int(os.environ.get("SLURM_CPUS_PER_TASK", os.cpu_count()))

# variables initialization
inputfile = "input.dat"         # input file
t_f = None                      # evolution final time
dt = None                       # time step 
NUMBER_OF_TRAJECTORIES = None   # number of trajectories to evolve
instate = None                  # initial state as a single character variable describing the evolution initial state {p (pure), m (maximally mixed)}
rho0 = None                     # initial state as a density matrix
alpha_over_kappa = None         # driving field intensity over the system emission rate
eta = None                      # detection efficiency
chunk_dim = None                # chunk dimension (number of trajectories to evolve simultaneously)

# we pass alpha/kappa and eta as arguments of command line
alpha_over_kappa = float(sys.argv[1])
eta = float(sys.argv[2])

# directory storing data and results
if eta == 0.4 and alpha_over_kappa == 1.0:
    directory = "figure1/"
elif eta == 1.0 and alpha_over_kappa == 0.4:
    directory = "figure2/"
else:
    directory = "./"

# reading from file simulation parameters
with open(directory + inputfile, "r") as f:
    # cycle on every line of the file
    for line in f:
        # words separated by a space within a line become the elements of a list
        parts = line.split()
        # it skips empty lines, controls that there are exactly two elements per line (otherwise it skips the line) and skips the comments
        if not line or len(parts) != 2 or line.startswith("#"):
            continue
        key, value = parts
        if key == "INSTATE":
            instate = value  
        elif key == "FINALT":
            t_f = float(value)
        elif key == "dt":
            dt = float(value)
        elif key == "NTRAJ":
            NUMBER_OF_TRAJECTORIES = int(value)
        elif key == "CHUNKDIM":
            chunk_dim = int(value)

# initial state as density matrix
if instate == "p":
    rho0 = ebasis[0] * ebasis[0].dag()  # ground state
elif instate == "m":
    rho0 = 0.5 * qeye(DIMH)             # maximally mixed state (one half the identity matrix)
else:
    raise ValueError("The initial state must be pure (p) or maximally mixed (m).")

process = "pd_" + instate + "_eta" + str(eta) + "_alpha" + str(alpha_over_kappa)    # process name
NUMBER_OF_TIMEINTERVALS = int(t_f / dt)                     # number of time intervals
tlist = np.linspace(0, t_f, NUMBER_OF_TIMEINTERVALS + 1)    # list of time intervals

print("System evolution (initial ", instate, " state, alpha/kappa = ", alpha_over_kappa, ", eta = ", eta, ", ", NUMBER_OF_TIMEINTERVALS, " time intervals and ", NUMBER_OF_TRAJECTORIES, " trajectories)...", sep = "")

# system operators
c = sigmam()    # collapse operator
# kraus operators
M0 = qeye(DIMH) - 1j * HS(alpha_over_kappa) * dt - 0.5 * c.dag() * c * dt   # no jump
M1 = np.sqrt(eta) * c * np.sqrt(dt)                                         # jump
# conditional jump probability
def P(rhot):
    return (eta * (c.dag() * c * rhot).tr() * dt).real
# function that returns the state evolution of a system subject to a contiuous photo-detection
def smesolve_kraus(rho_start):
    
    rhot = rho_start    # initial state
    results = [rhot]
    # system evolution
    for i in range(1, NUMBER_OF_TIMEINTERVALS):
        pjump = P(rhot)     # jump probability
        r = rnd.random()
        kraus = None        # kraus operator
        if r <= pjump: kraus = M1   # if a jump occurs
        else: kraus = M0            # if not
        num = kraus * rhot * kraus.dag() + (1 - eta) * c * rhot * c.dag() * dt
        rhot = num / num.tr()       # successive conditional state
        results.append(rhot)
    
    return results

print("System evolution (initial ", instate, " state, alpha/kappa = ", alpha_over_kappa, ", eta = ", eta, ", ", NUMBER_OF_TIMEINTERVALS, " time intervals and ", NUMBER_OF_TRAJECTORIES, " trajectories)...", sep = "")

int_time = 0    # progressive run time
start = time.time()
chunk_num = 0   # chunk number
with ProcessPoolExecutor(max_workers = num_cores) as executor:
    for i in range(0, NUMBER_OF_TRAJECTORIES, chunk_dim):
        chunk_num += 1
        states = []
        chunk_start_time = time.time()  # we start counting the execution time of the chunk
        # we map in parallel the stochastic Kraus master equation solver at chunk of initial states with dimension chunk_dim
        states = list(executor.map(smesolve_kraus, [rho0] * (chunk_dim)))
        # we append the states on a file
        with open("states/" + process + "_chunk" + str(chunk_num) + ".pkl", "ab") as f:
            pickle.dump(states, f)      # we print on a file the states evolutions of each chunk
        del states          # remove reference to solver result to free memory
        gc.collect()        # force garbage collection to release unused memory
        chunk_end_time = time.time()    # we end counting the execution time of the chunk
        int_time += chunk_end_time - chunk_start_time
        print(round(int(i + chunk_dim) / NUMBER_OF_TRAJECTORIES * 100, 1), "%. Run time: ", round(int_time, 2), "s.", sep = "")

end = time.time()

# total execution time
print("Total run time: ", round(end - start, 2), "s", sep = "")

File: code/photodet.jl
# === DYNAMICS SIMULATION OF A SYSTEM SUBJECT TO A CONTINUOUS PHOTO-DETECTION ===

# including required libraries
using Distributed   # for parallel computing

# no arguments reading from terminal

# variables initialization
inputfile = "input.dat"             # name of the file from which we read the simulation's paremeters
instate = nothing                   # single character variable that indicates the simulation's initial state 
α_val = nothing                     # resonant field intensity over emitting rate value
η_val = nothing                     # detectiong efficiency value
t_f = nothing                       # simulation's final time
deltat = nothing                    # simulation's time step
NUMBER_OF_TRAJECTORIES = nothing    # simulation's number of trajectories
chunk_dim = nothing                 # number of trajectories to evolve simultaneously

# reading the remaining parameters from file
for line in eachline(inputfile)
    # to split line's elements
    parts = split(line)
    # conditions to skip a line
    if isempty(line) || length(parts) != 2 || startswith(line, "#")
        continue
    end
    key, value = parts
    if key == "INSTATE"
        # "global" indicates a global variable
        global instate = value
    elseif key == "ALPHA"
        global α_val = parse(Float64, value)
    elseif key == "ETA"
        global η_val = parse(Float64, value)
    elseif key == "FINALT"
        global t_f = parse(Float64, value)
    elseif key == "dt"
        global deltat = parse(Float64, value)
    elseif key == "NTRAJ"
        global NUMBER_OF_TRAJECTORIES = parse(Int64, value)
    elseif key == "CHUNKDIM"
        global chunk_dim = parse(Int64, value)
    end
end

# initial state density matrix definition
if instate == "p"
    global ρ_0 = ComplexF64[0.00000000 0.00000000 ; 0.00000000 1.00000000]  # ground state
elseif instate == "m"
    global ρ_0 = ComplexF64[0.50000000 0.00000000 ; 0.00000000 0.50000000]  # maximally mixed state
else
    error("The initial state must be pure (p) or maximally mixed (m).")
end

# check on detection efficiency value
if η_val < 0 || η_val > 1
    error("The detection efficiency must be between 0 and 1.")
end

# string that identifies the input simulation's parameters
inputstring = instate * "_eta" * string(η_val) * "_alpha" * string(α_val)
# string that identifies process name
process = "pd_" * inputstring
# path where to save the states' trajectories
processpath = "states/" * inputstring * "/"
mkpath(processpath)
# write on processpath the number of trajectories
write(joinpath(processpath, "ntraj.dat"), string(NUMBER_OF_TRAJECTORIES), "\n")

# workers initialization
@everywhere begin
    # libraries inclusion for each worker
    using JLD2  # necessary because @save is called inside each worker
    include("my_library/my_objects.jl")
    
    # constants definition for each core
    α_over_κ = $α_val
    η = $η_val
    c = σ_m         # collapse operator
    finalt = $t_f
    dt = $deltat
    
    NUMBER_OF_TIMEINTERVALS = Int64(finalt / dt)
    tlist = range(0, finalt, NUMBER_OF_TIMEINTERVALS + 1)

    # pevolution function definition
    function pevolution(ρ_0)
        ρ_t = ρ_0   # initial state at time t=0
        results = [ρ_t]
        for i in tlist
            ρ_tdt = photodet_kraus(HS(α_over_κ), ρ_t, c, η)
            push!(results, ρ_tdt)
            ρ_t = ρ_tdt
        end
        return results
    end
end

println("System evolution (initial ", instate, " state, α/κ = ", α_val, ", η = ", η_val, ", ", NUMBER_OF_TIMEINTERVALS, " time intervals and ", NUMBER_OF_TRAJECTORIES, " number of trajectories)...")

prog_time = 0                                           # progressive run time
start_time = time()                                     # initial run time
chunk_ind = 0                                           # to count the chunk number
chunk_num = Int64(NUMBER_OF_TRAJECTORIES / chunk_dim)   # number of chunk

# we compute how many trajectories within a chunck are up to each worker
traj_per_worker = div(chunk_dim , nworkers())
# division remainder (it is possible that the number of trajectory per chunk is not a multiple of the number of workers)
rem = chunk_dim % nworkers()

for i in 1:chunk_num
    global chunk_ind += 1
    chunk_start_time = time()  
    # we prepare the groups of trajectories per worker
    # sync: wait for each worker to finish its task
    @sync begin
        # cycle over workers
        for (w_idx, w) in enumerate(workers())
            # we asign an extra trajectory to the first 'rem' workers
            n_per_worker = traj_per_worker + (w_idx <= rem ? 1 : 0)
            # remotecall assigns a task to a specific worker
            @async remotecall_fetch(w, ρ_0, chunk_ind, w_idx) do rho, c_id, w_id    # variable redefinition
                
                # each worker locally executes its sub-chunk
                local_states = [pevolution(rho) for j in 1:n_per_worker]
                
                # each worker saves its local results on disk
                filename = processpath * "$(process)_chunk$(c_id)_worker$(w_id).jld2"
                @save filename local_states
            end
        end
    end
    
    chunk_end_time = time()
    global prog_time += chunk_end_time - chunk_start_time
    println(round(Int64(i * chunk_dim) / NUMBER_OF_TRAJECTORIES * 100, digits = 1), "%. Run time: ", round(prog_time, digits = 2), "s.")
end

end_time = time()
println("Total run time: ", round(end_time - start_time, digits = 2), "s.")
